### PR TITLE
Slam 55 - Adding stereo cam

### LINKED ISSUE #55: Simulate ZED Stereo Camera
**Description**:
- Simulate ZED Stereo Cam in gazebo
- There are no ZED plugins, so need to emulate it

**Goals**:
- [x] Stereo camera gazebo plugin
- [x] Stereo camera frame in URDF
- [x] Emulate ZED depth map
- [x] Emulate ZED visual odometry

**Acceptance Criteria**:
- [x] Emulated ZED Stereo camera publishes data in simulation

**Notes**:
- Interface the simulated sensor with any packages that require it
- For ZED published topics see: https://www.stereolabs.com/blog/use-your-zed-camera-with-ros/

### PR DESCRIPTION
**Pull request by Federico-PizarroBejarano**
Added stereo cam, disparity, point cloud, and depth.

### CODE REVIEW

File: slam-ws/src/motor_control/launch/motor_control.launch [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <launch>
 
   <!-- Ports -->
-	<arg name="motor_port" default="/dev/ttyACM0"/>
+    <arg name="motor_port" default="/dev/ttyACM0"/>
   <arg name="encoder1_port" default="/dev/ttyACM1"/>
   <arg name="encoder2_port" default="/dev/ttyACM2"/>
```
Comment (erickmu1) [inline slam-ws/src/motor_control/launch/motor_control.launch:6]: why did the indent change?

File: slam-ws/src/igvc_world/models/igvc_basic_ground/materials/scripts/igvc_basic_ground.material [hunk 2]
```diff
@@ -1,7 +1,7 @@
 material igvc_basic_ground/Image
 {
-	technique
-	{
+    technique
```
Comment (erickmu1) [inline slam-ws/src/igvc_world/models/igvc_basic_ground/materials/scripts/igvc_basic_ground.material:3]: there are weird indent changes in this file

File: slam-ws/src/caffeine/urdf/inertias.xacro [hunk 3]
```diff
@@ -1,25 +1,25 @@
...
+    <xacro:macro name="cylinder_inertia" params="mass length radius">
+        <inertia
+            ixx="${mass * (3*radius*radius + length*length)/12.0}" 
+            ixy="0.0" 
+            ixz="0.0" 
                         iyy="${mass * (3*radius*radius + length*length)/12.0}" 
-			iyz="0.0" 
-			izz="${0.5*mass*radius*radius}"
-		/>
-	</xacro:macro>
```
Comment (erickmu1) [inline slam-ws/src/caffeine/urdf/inertias.xacro:24]: there are weird indent changes in this file

File: slam-ws/src/caffeine/launch/caffeine_setup.launch [hunk 4]
```diff
@@ -28,7 +28,7 @@
         <param name="serial_port" value="$(arg lidar_port)"/>
     </node>
 
-	<!-- Add Lidar Filtering -->
```
Comment (erickmu1) [inline slam-ws/src/caffeine/launch/caffeine_setup.launch:31]: there are weird indent changes in this file

File: slam-ws/src/motor_control/src/cmd_vel_to_motor.cpp [hunk 5]
```diff
@@ -53,7 +53,7 @@ int main(int argc, char **argv)
     ros::Rate loop_rate(100);
 
     std_msgs::Float64 rvelmsg;
-    std_msgs::Float64 lvelmsg;
```
Comment (erickmu1) [inline slam-ws/src/motor_control/src/cmd_vel_to_motor.cpp:56]: there are weird space changes in this file

File: slam-ws/src/rosserial_python/src/rosserial_python/SerialClient.py [hunk 6]
```diff
@@ -241,7 +241,7 @@ def listen(self):
             self.socket = clientsocket
             self.isConnected = True
 
-            if (self.fork_server == True):	# if configured to launch server in a separate process
```
Comment (erickmu1) [inline slam-ws/src/rosserial_python/src/rosserial_python/SerialClient.py:244]: there are weird space changes in this file

**Comment by erickmu1** [comment]
Can you give the topics the correct name according to the namespace `/zed/zed_node/`. See [here](https://www.stereolabs.com/docs/ros/) for the names of the topics you published. Also idk if you need rectified images from the stereo cam... You can also move this into the simulate ZED stereo cam issue

File: slam-ws/src/motor_control/launch/motor_control.launch (hunk 1 above)
Comment (Federico-PizarroBejarano) [reply slam-ws/src/motor_control/launch/motor_control.launch:6]: I switched all the tabs in all our files to 4 spaces

**Comment by Federico-PizarroBejarano** [comment]
> Can you give the topics the correct name according to the namespace `/zed/zed_node/`. See [here](https://www.stereolabs.com/docs/ros/) for the names of the topics you published. Also idk if you need rectified images from the stereo cam... You can also move this into the simulate ZED stereo cam issue

So the new namespace needs to be /zed/zed_node? where does it say that? The rectified images are produced automatically.

**Comment by Federico-PizarroBejarano** [comment]
@erickmu1 I added the visual odom package and added it to the gazebo launch file. the topics are created and seem to be working but I don't know how to test it

**Comment by erickmu1** [comment]
> > Can you give the topics the correct name according to the namespace `/zed/zed_node/`. See [here](https://www.stereolabs.com/docs/ros/) for the names of the topics you published. Also idk if you need rectified images from the stereo cam... You can also move this into the simulate ZED stereo cam issue
> 
> So the new namespace needs to be /zed/zed_node? where does it say that? The rectified images are produced automatically.

this is actually happens when you run the launch files they provide, all the topics are namespaced by `/zed` and it seems there is another layer of namespacing under `/zed_node`. Hence when you launch the stereo cam, you see topics such as `/odom` being published under `/zed/zed_node/odom`. Really this is just how it is when we launch stuff and would make simulation mimic reality a little more (plus less worrying over changing topic names).

**Comment by erickmu1** [comment]
> @erickmu1 I added the visual odom package and added it to the gazebo launch file. the topics are created and seem to be working but I don't know how to test it

You can move the robot and see if the odom readings match the direction you are moving in. Furthermore, you could compare the values between visual_odom and `odom/filtered` and see if they are about the same/consistent when the robot is moving. To move the robot u could either give a nav goal or use teleop keyboard, I'll leave it up to you

**Pull request merged**